compat/win32/dirent.h

```diff
@@ -8,7 +8,7 @@
 
 struct dirent {
 	unsigned char d_type; /* file type to prevent lstat after readdir */
-	char d_name[FLEX_ARRAY]; /* file name */
+	char d_name[MAX_PATH * 3]; /* file name (x3 for UTF-8 conversion) */
```

Comment:
Something about this bothers me.  Granted, I've spent about 10 minutes looking at the code, so I may need more education....

Upstream there was a d_name[MAX_PATH * 3] in "0217569bb2 (Win32: Unicode file name support (dirent), 2012-01-14)"

In git-for-windows, that got changed (or independently evolved) into d_name[FLEX_ARRAY] in "40e4213855 (mingw: make the dirent implementation pluggable, 2013-09-08)".  And it includes a `dir = xmalloc(sizeof(dirent_DIR) + MAX_PATH);`

However, that malloc was changed in "64da6f2182 (mingw: support long paths, 2015-07-28)
" to `dir = xmalloc(sizeof(dirent_DIR) + MAX_LONG_PATH);`

So I wonder if this should be `char d_name[MAX_LONG_PATH * 3]` or something?
Or maybe we need to revisit (if possible) the places where we put a `struct dirent` inside something else.

Again, I only spent a few minutes looking at this, so I may be off-base here.

Reply:
It is quite unclear to me how the claim "this runs successful all tests, but it is not efficient as the allocation could be tightened further" could be backed up with evidence. There is a single `xmalloc()`, no matter how you hold the code.

Reply:
Right. My bad.  256 per filename component. Virtually unlimited components.

Also, if this is intended for upstream, we don't have the MAX_LONG_PATH "long path" commits up there yet.

Reply:
Note that in Linux (and other systems), the d_name is "practically" a FLEX_ARRAY, because there is a an uint16_t d_namlen entry that tells what the "real" length of that name is, but that field is missing from this implementation and hence I agree with Jeff, this code looks fishy overall and besides the fact that it relies in undefined behaviour.

Reply:
> it relies in undefined behaviour.

Care to elaborate?

Reply:
FWIW there was a proposal to allow this as part of C17 which didn't make it into the standard, and it doesn't seem to be part of the last draft of C2X either:

  http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2083.htm

But synthetic tests for gcc and MSVC (both with cl and clang) show they all support this as an extension (even if it might trigger warnings) and more importantly are consistent on the memory layout used.

Reply:
Where do we use the flex array-containing struct within another struct? I may miss something blatantly obvious here, but if not, there is simply no undefined behavior. Color me still confused by your statements.